peer.go

```diff
@@ -1,516 +1,518 @@
-// Use of this source code is governed by an ISC
-// license that can be found in the LICENSE file.
-
 package main
 
 import (
-	"bytes"
-	"container/list"
 	"fmt"
-	"io"
-	prand "math/rand"
-	"net"
-	"strconv"
-	"sync"
-	"sync/atomic"
 	"time"
 
 	"github.com/btcsuite/btcd/addrmgr"
-	"github.com/btcsuite/btcd/blockchain"
 	"github.com/btcsuite/btcd/database"
+	"github.com/btcsuite/btcd/peer"
 	"github.com/btcsuite/btcd/wire"
 	"github.com/btcsuite/btcutil"
 	"github.com/btcsuite/btcutil/bloom"
-	"github.com/btcsuite/go-socks/socks"
-	"github.com/davecgh/go-spew/spew"
-)
-
-const (
-	// maxProtocolVersion is the max protocol version the peer supports.
-	maxProtocolVersion = 70002
-
-	// outputBufferSize is the number of elements the output channels use.
-	outputBufferSize = 50
-
-	// invTrickleSize is the maximum amount of inventory to send in a single
-	// message when trickling inventory to remote peers.
-	maxInvTrickleSize = 1000
-
-	// maxKnownInventory is the maximum number of items to keep in the known
-	// inventory cache.
-	maxKnownInventory = 1000
-
-	// negotiateTimeoutSeconds is the number of seconds of inactivity before
-	// we timeout a peer that hasn't completed the initial version
-	// negotiation.
-	negotiateTimeoutSeconds = 30
-
-	// idleTimeoutMinutes is the number of minutes of inactivity before
-	// we time out a peer.
-	idleTimeoutMinutes = 5
-
-	// pingTimeoutMinutes is the number of minutes since we last sent a
-	// message requiring a reply before we will ping a host.
-	pingTimeoutMinutes = 2
-
-	// connectionRetryInterval is the base amount of time to wait in between
-	// retries when connecting to persistent peers.  It is adjusted by the
-	// number of retries such that there is a retry backoff.
-	connectionRetryInterval = time.Second * 10
-
-	// maxConnectionRetryInterval is the max amount of time retrying of a
-	// persistent peer is allowed to grow to.  This is necessary since the
-	// retry logic uses a backoff mechanism which increases the interval
-	// base done the number of retries that have been done.
-	maxConnectionRetryInterval = time.Minute * 5
-)
-
-var (
-	// nodeCount is the total number of peer connections made since startup
-	// and is used to assign an id to a peer.
-	nodeCount int32
-
-	// userAgentName is the user agent name and is used to help identify
-	// ourselves to other bitcoin peers.
-	userAgentName = "btcd"
-
-	// userAgentVersion is the user agent version and is used to help
-	// identify ourselves to other bitcoin peers.
-	userAgentVersion = fmt.Sprintf("%d.%d.%d", appMajor, appMinor, appPatch)
 )
 
-// zeroHash is the zero value hash (all zeros).  It is defined as a convenience.
-var zeroHash wire.ShaHash
+// handleVersionMsg returns a handler which is invoked when a peer
+// receives a version bitcoin message and is used to negotiate the
+// protocol version details as well as kick start the communications.
+func handleVersionMsg(s *server) func(p *peer.Peer, msg *wire.MsgVersion) {
```

Comment:
I thought it might be useful to allow composition to provide any other context that might be required later, e.g like in: https://gist.github.com/tsenart/5fc18c659814c078378d

Although I agree that it might be simpler to keep them on the `Server`, will check and refactor that.